### PR TITLE
Add various methods to configure default plugins

### LINKED ISSUE #15: Document how javascript is being minified in production
I can't find how javascript is being minified in production or what additional options we can give it (e.g. stripping special comments).

### LINKED ISSUE #87: Make clean-webpack-plugin configurable
From a tweet! 
https://twitter.com/DevCircus/status/884574052127518720

### LINKED ISSUE #148: Expose mangle option for UglifyJsPlugin
Greetings!

Came across the problem of AgularJs minification. Some third party modules are not minified correctly with mangle option enabled. It throws error like ``Error: Unknown provider: tProvider <- t``. Unfortunately, there is no way to disable mangle via Encore interface. Please, add the option. Thank you in advance!

P.S. For those, experiencing the same problem, here's how I solved it:

```js
...

let config = Encore.getWebpackConfig(),
    webpack = require('webpack');

if (Encore.isProduction()) {
    let index;

    config.plugins.forEach((plugin, i) => {
        if (plugin instanceof webpack.optimize.UglifyJsPlugin) {
            index = i;
        }
    });

    if (index) {
        config.plugins.splice(index, 1);
    }

    config.plugins.push(new webpack.optimize.UglifyJsPlugin({
        mangle: false,
        sourceMap: !Encore.isProduction()
    }));
}
```

### PR DESCRIPTION
**Pull request by Lyrkan**
This PR adds some methods to configure default plugins (closes #148, closes #87 and closes #15):

* `Encore.configureDefinePlugin(callback)`
* `Encore.configureExtractTextPlugin(callback)`
* `Encore.configureFriendlyErrorsPlugin(callback)`
* `Encore.configureLoaderOptionsPlugin(callback)`
* `Encore.configureUglifyJsPlugin(callback)`

Other changes:

* Allows the `clean-webpack-plugin` to be configured using `Encore.cleanupOutputBeforeBuild(paths, callback)`
* Fixed a small mistake in the `Encore.configureManifestPlugin()` jsdoc
* Reorganized flags/callbacks in the constructor of `webpack.config.js` since it was starting to be a bit hard to read. I'm not sure about the way I splitted things though, let me know what you think about it.
* Renamed `common-chunks.js` to `commons-chunks.js` since the name of the plugin is `CommonsChunkPlugin`

@weaverryan: Not directly related but while doing all of that I noticed that `sassOptions` uses snake-case whereas I used camel-case in #144 for `enablePreactPreset()` and in #115 for `configureRuntimeEnvironment()`, should we do something about that?

***Edit 1:** Added `Encore.configureCleanWebpackPlugin()`*
***Edit 2:** Removed `Encore.configureCleanWebpackPlugin()` to use `Encore.cleanupOutputBeforeBuild(paths, callback)` arguments instead*

### CODE REVIEW

**Comment by weaverryan** [comment]
> Not directly related but while doing all of that I noticed that sassOptions uses snake-case whereas I used camel-case in #144 for enablePreactPreset() and in #115 for configureRuntimeEnvironment(), should we do something about that?

Hmm, yea, good catch. We should probably "get this right" now. I'm leaning towards using camelCase... so we would "fix" `sassOptions` (I'm leaning towards this in part because the options in #115 are camelCase just because that's how the CLI options are normalized internally.

File: index.js [hunk 1]
```diff
@@ -103,13 +103,89 @@ const publicApi = {
...
+     * @param {function} loaderOptionsPluginOptionsCallback
+     * @returns {exports}
+     */
+    configureLoaderOptionsPlugin(loaderOptionsPluginOptionsCallback = () => {}) {
+        webpackConfig.configureLoaderOptionsPlugin(loaderOptionsPluginOptionsCallback);
+
+        return this;
+    },
+
+    /**
```
Comment (weaverryan) [inline index.js]: This all looks beautiful

File: index.js [hunk 2]
```diff
@@ -522,10 +618,20 @@ const publicApi = {
      * If enabled, the output directory is emptied between
      * each build (to remove old files).
```
Comment (weaverryan) [inline index.js]: It's not from your commit, but this could probably fit all on one line.

File: index.js [hunk 3]
```diff
@@ -522,10 +618,20 @@ const publicApi = {
...
      *
+     * A list of available options can be found at https://github.com/johnagan/clean-webpack-plugin
+     *
+     * For example:
+     *
+     *      Encore.cleanupOutputBeforeBuild(['*.js'], (options) => {
```
Comment (weaverryan) [inline index.js]: Maybe we should show the default `['**/*']` as the first option. The issue is that, if I *only* want to specify some options for the plugin... I need to know what this first value is by default, so I can repeat it. That kind of sucks... but I can't think of a good way around it (so showing the default here might make sense)

File: lib/WebpackConfig.js [hunk 4]
```diff
@@ -311,8 +371,18 @@ class WebpackConfig {
         this.configuredFilenames = configuredFilenames;
     }
 
-    cleanupOutputBeforeBuild() {
+    cleanupOutputBeforeBuild(paths = ['**/*'], cleanWebpackPluginOptionsCallback = () => {}) {
+        if (!Array.isArray(paths)) {
+            throw new Error('Argument 1 to cleanupOutputBeforeBuild() must be an Array');
```
Comment (weaverryan) [inline lib/WebpackConfig.js]: What do you think about this:

> ... must be an Array of paths - e.g. ['**/*']

File: index.js (hunk 3 above)
Comment (Lyrkan) [reply index.js]: I wanted to but `'**/*'` breaks the multi-line comment :(

Comment (weaverryan) [reply index.js]: Ha! What a crazy edge case :p

**Comment by weaverryan** [comment]
Thank you @Lyrkan!

**Pull request merged**